File: Leetcode/js/29_divideTwoIntegers.js
/*
Divide two integers without using multiplication, division and mod operator.

If it is overflow, return MAX_INT.
*/

/**
 * @param {number} dividend
 * @param {number} divisor
 * @return {number}
 */
var divide = function(dividend, divisor) {
    if(divisor === 0){
        return 0;
    }
    
    var max32 = (Math.pow(2,31)-1);
    var min32 = -Math.pow(2,31);
    
    var multiplier = 1;
    var binser = function() {
        while(dividend - (divisor*multiplier*2) >= 0 ) {
            multiplier *= 2;
        }
    };
    
    var result = 0;
    var neg = false;
    
    if(dividend<0){
        neg = !neg;
        dividend = -1*dividend;
    }
    if(divisor<0){
        neg = !neg;
        divisor = -1*divisor;
    }
    
    binser();
    while(dividend >= divisor){
        while(dividend-(divisor*multiplier) >= 0) {
            result += multiplier;
            dividend -= (divisor*multiplier);
        }
        if(multiplier>1) multiplier /= 2;
    }
    
    return neg ? Math.max(-1*result, min32) : Math.min(result, max32);
};

File: Leetcode/java/29_divideTwoIntegers.java
/*
Divide two integers without using multiplication, division and mod operator.

If it is overflow, return MAX_INT.
*/


class Main {
    public static void main(String[] args) {
        System.out.println("Starting...");

        Solution solution = new Solution();

        int input = -1;
        int input2 = 1;

        int k = 1;
        
        int result = solution.divide(input, input2);
        
        System.out.println("Result: " + result);
        
        System.out.println("DONE!");
    }
}

class Solution {
    public int divide(int dividend, int divisor) {
        if(divisor == 0) {
            throw new IllegalArgumentException("Div by zero");
        }

        boolean signNeg = dividend < 0 ^ divisor < 0;

        long ldend = Math.abs((long) dividend);
        long lsor = Math.abs((long) divisor);

        long result = 0;
        int multiplier = 0;

        while(ldend > lsor << multiplier+1) {
            multiplier++;
        }

        long currentDivisor;

        while(ldend >= lsor) {
            currentDivisor = lsor << multiplier;
            
            if(ldend >= currentDivisor) {
                ldend -= currentDivisor;
                result += 1 << multiplier;
            } else {
                multiplier--;
            }
        }
        
        if(!signNeg && result > Integer.MAX_VALUE) {
            return Integer.MAX_VALUE;
        }

        if(signNeg && result-1 > Integer.MAX_VALUE) {
            return Integer.MIN_VALUE;
        }

        return signNeg ? (int)-result : (int)result;

    }
}






